### PR TITLE
#162948989 Descriptive Registration validation errors

### PR DESCRIPTION
**Pull request by Hesbon5600**
**What does this PR do?**

_- Email must be valid and unique_
_- Password must be alphanumeric_
_- Password must have an uppercase and lowercase letter_
_- Password must be at least 8 characters long_
_- Username must be unique_

**How should this be manually tested?**
prerequisites:

Have postgres installed and running on localhost.

Follow the following steps:
1. Clone this repo with the following command:
`https://github.com/andela/ah-fulldeck.git`

1. Cd into the ah-fulldeck directory:
`cd ah-fulldeck`

1. Checkout to the branch `ft-validate-signup-162948989` with the command:
`git checkout ft-validate-signup-162948989`

1. Install the requirements by running the command:
`pip install requirements.txt`

1. Create a database:
`create database fulldeck`

1. Export the environment settings as follows:
`export DATABASE="fulldeck"`
`export USER="your postgres username "`
`export HOST="localhost"`
`export PASSWORD="your postgres password"`

1. Migrate the models to the db:
 `python manage.py migrate`

1. Run the server:
`python manage.py runserver`

1. From postman:
Create a user by sending a post request to the '/api/user/' end point with the data body 
`{`
     `"user": {`
                  `"username" :"prefered name",`
                  `"email": "your email",`
                  `"password": "password"`
`}`
`}`
You can change the format of the email or password.
Invalid email and password will result in a response message outlinging the fault.
An existing username or email will also throw an error message

**Any background context you want to provide?**
None

**What are the relevant pivotal tracker stories?**
[#162948989](https://www.pivotaltracker.com/story/show/162948989)

**Screenshots**

1. Correct email and password format 
screenshot 2019-01-15 at 15 45 16
2. Invalid email format
screenshot 2019-01-15 at 15 45 57
3. Password without a number or letter
screenshot 2019-01-16 at 10 37 34

### CODE REVIEW

**Comment by eduhmik** [comment]
Hey @Hesbon5600 keep up the good spirit... however on your PR in the steps to follow when testing add an option to install virtualenv for the application and how to activate it on both OS

**Comment by charitymarani** [comment]
The error messages are descriptive. Good work.

**Comment by Hesbon5600** [comment]
Okay @edhmik I'll do that

File: authors/apps/authentication/serializers.py [hunk 1]
```diff
@@ -10,10 +10,53 @@ class RegistrationSerializer(serializers.ModelSerializer):
...
     # characters, and can not be read by the client.
-    password = serializers.CharField(
-        max_length=128,
+    # serializers.RegexField - A text representation, that validates the given value matches against a certain regular expression.
+
+    password = serializers.RegexField(
+        regex="^(?=.*[A-Za-z])(?=.*\d)[A-Za-z\d]{8,}$",
         min_length=8,
-        write_only=True
+        max_length=128,
```
Comment (kzyangiro) [inline authors/apps/authentication/serializers.py]: @Hesbon5600 128 is alot, reset this to 30 max characters.

Comment (Hesbon5600) [reply authors/apps/authentication/serializers.py]: Okay sure

File: authors/apps/authentication/serializers.py [hunk 2]
```diff
@@ -10,10 +10,53 @@ class RegistrationSerializer(serializers.ModelSerializer):
...
+        ],
+        error_messages={
+            'invalid': 'Email must be of the format name@domain.com',
+            'required': 'Email is a required field'
+        }
+    )
+    # The username must be: unique, at least 5 characters long
+    # There should not be any space in the username
+    # Underscores and hyphens are also allowed in the username
+
```
Comment (kzyangiro) [inline authors/apps/authentication/serializers.py:46]: at least 5? I thought in the implementation you used 8..

Comment (Hesbon5600) [reply authors/apps/authentication/serializers.py:46]: I think It should just be '# The username must be: unique'. Let me fix that. 
The password  is the one that must be at least 8 characters

**Review by kzyangiro — APPROVED** [review]

**Pull request merged**